File: triangsup.py
from turtle import end_fill
from numpy import zeros,array

def triangsup(T,b):
    #T est la matrice triangulaire supérieure
    #b est le vecteur second membre
    n=len(b)
    x=zeros(n)
    x[n-1]=b[n-1]/T[n-1][n-1]
    for i in range(n-2,-1,-1):
        S=0
        for k in range(i+1,n):
            S+=T[i][k]*x[k]
        x[i]=(b[i]-S)/T[i][i]
    return x

# x=triangsup(array([[4,0,0],[3,-3,0],[-1,1,1]]), array([8, 9, -2]))
# print(x)
    

File: julia/triangsup.jl
function triangSup(A::Matrix{T}, b::Vector{T}) where T <: Number
    n = size(A, 1)
    x = zeros(n)
    x[n] = b[n] / A[n, n]
    for i = n-1:-1:1
        Σ = 0
        for j = i+1:n
            Σ += A[i, j] * x[j]
        end
        x[i] = (b[i] - Σ) / A[i, i]
    end
    return x
end